go-controller/pkg/clustermanager/subnetallocator/host_subnet_allocator.go

```diff
@@ -3,57 +3,329 @@ package subnetallocator
+	}
+
+	if err := annotator.Set(hotypes.HybridOverlayNodeSubnet, hostSubnets[0].String()); err != nil {
+		if e := sna.hybridOverlaySubnetAllocator.ReleaseNetworks(node.Name, allocatedSubnets...); e != nil {
+			klog.Warningf("Failed to release hybrid over subnet for the node %s from the allocator : %w", node.Name, e)
+		}
+		return nil, fmt.Errorf("error setting hybrid overlay host subnet: %w", err)
+	}
+
+	return hostSubnets[0], nil
+}
+
+func (sna *HostSubnetAllocator) releaseHybridOverlayNodeSubnet(nodeName string) {
+	sna.hybridOverlaySubnetAllocator.ReleaseAllNetworks(nodeName)
+	klog.Infof("Deleted hybrid overlay HostSubnets for node %s", nodeName)
+}
+
+// HandleAddUpdateNodeEvent handles the add or update node event
+func (sna *HostSubnetAllocator) HandleAddUpdateNodeEvent(node *corev1.Node) error {
+	if util.NoHostSubnet(node) {
+		if sna.hasHybridOverlayAllocation() && houtil.IsHybridOverlayNode(node) {
+			annotator := kube.NewNodeAnnotator(sna.kube, node.Name)
+			allocatedSubnet, err := sna.hybridOverlayNodeEnsureSubnet(node, annotator)
+			if err != nil {
+				return fmt.Errorf("failed to update node %s hybrid overlay subnet annotation: %v", node.Name, err)
+			}
+			if err := annotator.Run(); err != nil {
+				// Release allocated subnet if any errors occurred
+				if allocatedSubnet != nil {
+					sna.releaseHybridOverlayNodeSubnet(node.Name)
+				}
+				return fmt.Errorf("failed to set hybrid overlay annotations for node %s: %v", node.Name, err)
+			}
+		}
+		return nil
+	}
+
+	err := sna.syncNodeNetworkAnnotations(node)
+	sna.recordSubnetUsage()
+	return err
+}
+
+// syncNodeNetworkAnnotations does 2 things
+//   - syncs the node's allocated subnets in the node subnet annotation
+//   - syncs the network id in the node network id annotation
+func (sna *HostSubnetAllocator) syncNodeNetworkAnnotations(node *corev1.Node) error {
+	networkName := sna.netInfo.GetNetworkName()
+
+	existingSubnets, err := util.ParseNodeHostSubnetAnnotation(node, networkName)
+	if err != nil && !util.IsAnnotationNotSetError(err) {
+		// Log the error and try to allocate new subnets
+		klog.Warningf("Failed to get node %s host subnets annotations for network %s : %v", node.Name, networkName, err)
+	}
+
+	networkID, err := util.ParseNetworkIDAnnotation(node, networkName)
+	if err != nil && !util.IsAnnotationNotSetError(err) {
+		// Log the error and try to allocate new subnets
+		klog.Warningf("Failed to get node %s network id annotations for network %s : %v", node.Name, networkName, err)
+	}
+
+	// On return validExistingSubnets will contain any valid subnets that
+	// were already assigned to the node. allocatedSubnets will contain
+	// any newly allocated subnets required to ensure that the node has one subnet
+	// from each enabled IP family.
+	ipv4Mode, ipv6Mode := sna.netInfo.IPMode()
+	validExistingSubnets, allocatedSubnets, err := sna.allocateNodeSubnets(sna.clusterSubnetAllocator, node.Name, existingSubnets, ipv4Mode, ipv6Mode)
```

Comment:
We don't use the info from sna.netInfo for hybrid overlay allocations, those are always ipv4 only. I guess if hybrid overlay is enabled, sna.netInfo should also have ipv6 disabled but it is being hardcoded for extra carefulness.

Reply:
I don't follow, sorry. 

I'm just asking if you can't implement 

```golang
func (sna *HostSubnetAllocator) allocateNodeSubnets(allocator SubnetAllocator, nodeName string, existingSubnets []*net.IPNet) ([]*net.IPNet, []*net.IPNet, error) {
    ipv4Mode, ipv6Mode := sna.netInfo.IPMode()
...
```

instead of 
```golang
func (sna *HostSubnetAllocator) allocateNodeSubnets(allocator SubnetAllocator, nodeName string, existingSubnets []*net.IPNet, ipv4Mode, ipv6Mode bool) ([]*net.IPNet, []*net.IPNet, error) {
```

Anyway, this is just a nit intending to avoid passing two extra arguments to the function.

Reply:
No because in line 113 is being called as
```
hostSubnets, allocatedSubnets, err := sna.allocateNodeSubnets(sna.hybridOverlaySubnetAllocator, node.Name, existingSubnets, true, false)
```
because of the reason I explained above.

Reply:
got it.